File: packages/sdk-py/aegis/key_storage.py
"""Pluggable persistence for Ed25519 private keys — Round 25 seed.

Mirror of the TS SDK's ``key-storage.ts``. The same threat model applies:
juniors dropping privateKey into a ``.env`` file leaks it via git; this
module gives them a sensible default that doesn't.

Implementations:

  - :func:`memory_key_storage`      — in-process dict. Lambda / tests.
  - :func:`file_system_key_storage` — ``~/.aegis/keys/`` mode 0600.
  - :class:`KmsKeyStorage`          — Protocol for remote signers (AWS
                                       KMS / GCP KMS / Vault). Provider
                                       implementations land in adapter
                                       packages, not the core SDK, to
                                       keep cloud deps optional.
"""
from __future__ import annotations

import json
import os
import re
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Optional, Protocol, runtime_checkable

from .runtime import capabilities

_VALID_NAME = re.compile(r"^[A-Za-z0-9_-]+$")


@dataclass(frozen=True)
class StoredKey:
    """One Ed25519 keypair record. ``private_key`` is base64url-encoded."""

    private_key: str
    """Base64url-encoded Ed25519 private key (32 bytes after decode)."""

    public_key: str
    """Base64url-encoded Ed25519 public key (32 bytes after decode)."""

    created_at: str
    """ISO 8601 timestamp of original creation."""

    agent_id: Optional[str] = None
    """Optional agentId once the key has been registered with AEGIS."""

    label: Optional[str] = None
    """Free-form human-readable label (matches ``AgentRecord.label``)."""


@runtime_checkable
class KeyStorage(Protocol):
    """Protocol every storage adapter must satisfy."""

    async def get(self, name: str) -> Optional[StoredKey]: ...
    async def put(self, name: str, key: StoredKey) -> None: ...
    async def delete(self, name: str) -> None: ...
    async def list(self) -> list[str]: ...


# ── In-memory ───────────────────────────────────────────────────────────────


@dataclass
class _MemoryKeyStorage:
    """In-process dict-backed storage. Lifetime is the Python process."""

    _store: dict[str, StoredKey] = field(default_factory=dict)

    async def get(self, name: str) -> Optional[StoredKey]:
        return self._store.get(name)

    async def put(self, name: str, key: StoredKey) -> None:
        self._store[name] = key

    async def delete(self, name: str) -> None:
        self._store.pop(name, None)

    async def list(self) -> list[str]:
        return list(self._store.keys())


def memory_key_storage() -> KeyStorage:
    """Build an in-memory KeyStorage. Use for tests and Lambda."""
    return _MemoryKeyStorage()


# ── Filesystem ──────────────────────────────────────────────────────────────


@dataclass
class _FileSystemKeyStorage:
    """Filesystem-backed storage. JSON files at ``<dir>/<name>.json``, mode 0600."""

    dir: Path

    def _file_for(self, name: str) -> Path:
        if not _VALID_NAME.match(name):
            raise ValueError(
                f'invalid key name "{name}" — allowed chars: [a-zA-Z0-9_-]',
            )
        return self.dir / f"{name}.json"

    async def get(self, name: str) -> Optional[StoredKey]:
        path = self._file_for(name)
        if not path.exists():
            return None
        data = json.loads(path.read_text("utf-8"))
        return StoredKey(**data)

    async def put(self, name: str, key: StoredKey) -> None:
        path = self._file_for(name)
        path.write_text(json.dumps(asdict(key)), encoding="utf-8")
        os.chmod(path, 0o600)

    async def delete(self, name: str) -> None:
        path = self._file_for(name)
        if path.exists():
            path.unlink()

    async def list(self) -> list[str]:
        if not self.dir.exists():
            return []
        return [p.stem for p in self.dir.glob("*.json")]


def file_system_key_storage(*, dir: Optional[Path] = None) -> KeyStorage:
    """Build a filesystem KeyStorage.

    Default directory is ``$AEGIS_KEY_DIR`` or ``~/.aegis/keys/`` (mode 0700).
    Refuses to construct on runtimes without a writable filesystem
    (e.g. AWS Lambda root FS).
    """
    caps = capabilities()
    if not caps.has_filesystem:
        raise RuntimeError(
            f"file_system_key_storage: runtime {caps.runtime!r} has no writable filesystem. "
            "Use memory_key_storage() or a KmsKeyStorage instead.",
        )
    target = dir or Path(os.environ.get("AEGIS_KEY_DIR", Path.home() / ".aegis" / "keys"))
    target.mkdir(parents=True, exist_ok=True, mode=0o700)
    return _FileSystemKeyStorage(dir=target)


# ── KMS adapter shape (Protocol only — providers in adapter packages) ────────


@runtime_checkable
class KmsKeyStorage(Protocol):
    """Marker Protocol for KMS-backed storage.

    The SDK never holds private bytes for KMS-backed keys; instead
    :meth:`sign` round-trips through the KMS Sign API. Provider
    implementations live in companion packages (``aegis_aws_kms``,
    ``aegis_gcp_kms``, ``aegis_vault``).
    """

    kind: str  # always 'kms'

    async def sign(self, name: str, message: bytes) -> str:
        """Sign ``message`` with the KMS-held private key. Returns base64url."""
        ...

    async def public_key(self, name: str) -> str:
        """Return the public key bytes. Never the private."""
        ...


def default_key_storage(*, dir: Optional[Path] = None) -> KeyStorage:
    """Pick the right default storage adapter for the current runtime.

    Returns :func:`file_system_key_storage` when the filesystem is
    writable, :func:`memory_key_storage` otherwise. Used by
    :func:`aegis.quickstart.quickstart` when the caller doesn't supply
    ``storage`` explicitly.
    """
    caps = capabilities()
    if caps.has_filesystem:
        return file_system_key_storage(dir=dir)
    return memory_key_storage()


__all__ = [
    "KeyStorage",
    "KmsKeyStorage",
    "StoredKey",
    "default_key_storage",
    "file_system_key_storage",
    "memory_key_storage",
]

File: packages/sdk-ts/src/key-storage.ts
// KeyStorage — Round 25 Lane A. Pluggable persistence for the Ed25519
// private keys that `generateKeypair()` produces.
//
// Why this exists:
//
// The SDK's README told developers "persist privateKey locally — AEGIS
// never receives it." Juniors then dropped the key into a `.env` file and
// leaked it via git, or into `localStorage` where any XSS could exfiltrate
// it. The CLAUDE.md invariant ("private keys never enter AEGIS") only
// holds when the developer's storage choice is itself sound.
//
// KeyStorage is the contract. Implementations:
//
//   - `memory()`        — in-process Map. Edge / serverless / tests.
//   - `fileSystem()`    — `~/.aegis/keys/` with 0600 perms. Node only.
//   - `indexedDB()`     — origin-scoped IndexedDB. Browser only.
//   - `kms({ provider })` — adapter shape for AWS/GCP/Vault KMS; never sees
//                            the raw private bytes, instead returns a signer
//                            that delegates to the KMS Sign API.
//
// All implementations encode the private key as base64url so the wire shape
// matches what `generateKeypair()` returns. Public keys travel alongside in
// the same record so the SDK can re-link an existing key to an agentId on
// reboot without a server round-trip.

import { capabilities } from './runtime.js';

export interface StoredKey {
  /** Base64url-encoded Ed25519 private key (32 bytes after decode). */
  privateKey: string;
  /** Base64url-encoded Ed25519 public key (32 bytes after decode). */
  publicKey: string;
  /** ISO timestamp of original creation. Set by the SDK on first put. */
  createdAt: string;
  /** Optional agentId once the key has been registered with AEGIS. */
  agentId?: string;
  /** Free-form label for human selection (matches AgentRecord.label). */
  label?: string;
}

export interface KeyStorage {
  /** Read a key by name. Returns undefined when absent. */
  get(name: string): Promise<StoredKey | undefined>;
  /** Write a key under a name. Overwrites silently. */
  put(name: string, key: StoredKey): Promise<void>;
  /** Remove a key. No-op when absent. */
  delete(name: string): Promise<void>;
  /** List all stored key names. Order is implementation-defined. */
  list(): Promise<string[]>;
}

// ── Implementations ─────────────────────────────────────────────────────────

/**
 * In-memory storage. Lifetime is the JS process. Use for tests, edge
 * workers (which restart per request anyway), and ephemeral CLI flows.
 */
export function memoryKeyStorage(): KeyStorage {
  const store = new Map<string, StoredKey>();
  return {
    async get(name) {
      return store.get(name);
    },
    async put(name, key) {
      store.set(name, key);
    },
    async delete(name) {
      store.delete(name);
    },
    async list() {
      return [...store.keys()];
    },
  };
}

/**
 * File-system storage. Writes to `$AEGIS_KEY_DIR` (or `~/.aegis/keys/` by
 * default). Each key is a JSON file at `<dir>/<name>.json` with mode 0600.
 * The directory itself is created with mode 0700 on first use.
 *
 * Refuses to run on non-Node runtimes (throws at construction). Use
 * `kms()` or `indexedDB()` instead in those environments.
 */
export function fileSystemKeyStorage(opts: { dir?: string } = {}): KeyStorage {
  const caps = capabilities();
  if (!caps.hasFilesystem) {
    throw new Error(
      `fileSystemKeyStorage: requires Node/Bun/Deno runtime (detected ${caps.runtime}). Use memoryKeyStorage() or indexedDBKeyStorage() instead.`,
    );
  }
  // Lazy-import node:fs / node:path / node:os so non-Node bundles don't
  // pull these in. The capability check above guards the dynamic require.
  // type-rationale: top-level `require` is unavailable in pure ESM; we
  // resolve via createRequire only on Node-shaped runtimes.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const req = (globalThis as any).require ?? eval('require');
  const fs = req('node:fs');
  const path = req('node:path');
  const os = req('node:os');

  const dir = opts.dir ?? path.join(os.homedir(), '.aegis', 'keys');
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true, mode: 0o700 });
  }
  const fileFor = (name: string): string => {
    // Strict allow-list: alphanumerics, dash, underscore. Keeps the key
    // name from escaping the directory via `../` or absolute paths.
    if (!/^[a-zA-Z0-9_-]+$/.test(name)) {
      throw new Error(`invalid key name "${name}" — allowed chars: [a-zA-Z0-9_-]`);
    }
    return path.join(dir, `${name}.json`);
  };
  return {
    async get(name) {
      const file = fileFor(name);
      if (!fs.existsSync(file)) return undefined;
      const raw = fs.readFileSync(file, 'utf8');
      return JSON.parse(raw) as StoredKey;
    },
    async put(name, key) {
      const file = fileFor(name);
      fs.writeFileSync(file, JSON.stringify(key), { mode: 0o600 });
    },
    async delete(name) {
      const file = fileFor(name);
      if (fs.existsSync(file)) fs.unlinkSync(file);
    },
    async list() {
      const entries: string[] = fs.readdirSync(dir);
      return entries
        .filter((e: string) => e.endsWith('.json'))
        .map((e: string) => e.replace(/\.json$/, ''));
    },
  };
}

/**
 * IndexedDB storage — browser only. Origin-scoped, persists across tabs
 * and reloads. Refuses on non-browser runtimes.
 *
 * NOTE: IndexedDB is plain-text storage. For high-stakes browser flows
 * the recommended pattern is `kms()` with a remote signer (the browser
 * never holds the raw private bytes). IndexedDB is for low-stakes /
 * development convenience.
 */
export function indexedDBKeyStorage(dbName = 'aegis-keys'): KeyStorage {
  const caps = capabilities();
  if (caps.runtime !== 'browser') {
    throw new Error(
      `indexedDBKeyStorage: requires browser runtime (detected ${caps.runtime}). Use kms() in non-browser environments.`,
    );
  }
  // type-rationale: globalThis.indexedDB is typed in lib.dom but TS sees
  // `any` because the runtime check above narrows differently than the
  // type system. We cast through `as any` once at the boundary.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const idb: IDBFactory = (globalThis as any).indexedDB;
  const STORE = 'keys';

  function open(): Promise<IDBDatabase> {
    return new Promise((resolve, reject) => {
      const req = idb.open(dbName, 1);
      req.onupgradeneeded = () => req.result.createObjectStore(STORE);
      req.onsuccess = () => resolve(req.result);
      req.onerror = () => reject(req.error ?? new Error('indexedDB open failed'));
    });
  }
  async function tx<T>(mode: IDBTransactionMode, op: (s: IDBObjectStore) => IDBRequest<T>): Promise<T> {
    const db = await open();
    return new Promise<T>((resolve, reject) => {
      const t = db.transaction(STORE, mode);
      const req = op(t.objectStore(STORE));
      req.onsuccess = () => resolve(req.result);
      req.onerror = () => reject(req.error ?? new Error('indexedDB op failed'));
    });
  }
  return {
    async get(name) {
      // type-rationale: get() returns `any` — narrow to StoredKey | undefined.
      const v = await tx<unknown>('readonly', (s) => s.get(name));
      return (v as StoredKey | undefined) ?? undefined;
    },
    async put(name, key) {
      await tx<IDBValidKey>('readwrite', (s) => s.put(key, name));
    },
    async delete(name) {
      await tx<undefined>('readwrite', (s) => s.delete(name));
    },
    async list() {
      const v = await tx<IDBValidKey[]>('readonly', (s) => s.getAllKeys());
      return v.map((k) => String(k));
    },
  };
}

// ── KMS adapter shape (no implementation, but typed) ────────────────────────

/**
 * Marker for KMS-backed storage. The SDK never holds private bytes; instead,
 * `sign()` round-trips through the KMS Sign API.
 *
 * Implementations live in companion adapter packages (`@aegis/adapter-aws-kms`,
 * etc.) — kept out of the core SDK to avoid pulling cloud SDKs into edge bundles.
 *
 * The Round 25 SDK ships the SHAPE; full provider implementations land in
 * Round 26 alongside `@aegis/adapter-aws-lambda` / `-vercel-edge` / etc.
 */
export interface KmsKeyStorage {
  readonly kind: 'kms';
  /** Sign a message with the KMS-held private key. Returns base64url. */
  sign(name: string, message: Uint8Array): Promise<string>;
  /** Return the public key bytes; never returns the private. */
  publicKey(name: string): Promise<string>;
}

/**
 * Pick the default storage adapter for the current runtime. Used by
 * `Aegis.quickstart()` when the caller didn't supply one explicitly.
 *
 *   - node / bun / deno  → fileSystem()
 *   - browser            → indexedDB()
 *   - edge / workers     → memory() (ephemeral; warns)
 */
export function defaultKeyStorage(opts: { dir?: string } = {}): KeyStorage {
  const caps = capabilities();
  if (caps.hasFilesystem) return fileSystemKeyStorage(opts);
  if (caps.runtime === 'browser') return indexedDBKeyStorage();
  return memoryKeyStorage();
}
